src/gui/sdl_mapper.cpp

```diff
@@ -2982,15 +2983,24 @@ void MAPPER_LosingFocus() {
 
 void MAPPER_RunEvent(uint32_t /*val*/)
 {
-	KEYBOARD_ClrBuffer();           // Clear buffer
-	GFX_LosingFocus();		//Release any keys pressed (buffer gets filled again).
+	// Clear buffer
+	KEYBOARD_ClrBuffer();           
+	
+	// Release any keys pressed (buffer gets filled again).
+	GFX_LosingFocus();		
+
 	MAPPER_DisplayUI();
 }
 
 void MAPPER_Run(bool pressed) {
-	if (pressed)
+	if (pressed) {
 		return;
-	PIC_AddEvent(MAPPER_RunEvent,0);	//In case mapper deletes the key object that ran it
+	}
+
+	MIXER_LockMixerThread();
```

Comment:
I don't like the lock and unlock not being in the same function.  Unlock being inside `MAPPER_DisplayUI()`.

There's actually undefined behavior here as well.  If you start Dosbox with `--startmapper`, it calls `MAPPER_DisplayUI()` directly.  In that case, we unlock a mutex while we're not holding it.

I'd recommend either putting the lock at the top of `MAPPER_DisplayUI()` or putting both the lock and unlock at the call site of `MAPPER_RunEvent`.  The latter doesn't stop the mixer for the `--startmapper` case but I don't think we would ever be playing sound there anyway at first launch.

Reply:
Good catch. I was quite tired when I did this yesterday night.

I put them into `MAPPER_DisplayUI()` because that's where the mouse & gfx disengage/engage calls are, so at least they're in one place. It's a royal mess, only the full rewrite will make this file better...